src/main/java/io/nats/client/impl/NatsConnectionWriter.java

```diff
@@ -75,17 +77,30 @@ enum Mode {
         int sbl = bufferAllocSize(options.getBufferSize(), BUFFER_BLOCK_SIZE);
         sendBufferLength = new AtomicInteger(sbl);
         sendBuffer = new byte[sbl];
-
-        normalOutgoing = new MessageQueue(true,
-            options.getMaxMessagesInOutgoingQueue(),
-            options.isDiscardMessagesWhenOutgoingQueueFull(),
-            options.getRequestCleanupInterval(),
-            sourceWriter == null ? null : sourceWriter.normalOutgoing);
-
-        // The "reconnect" buffer contains internal messages, and we will keep it unlimited in size
-        reconnectOutgoing = new MessageQueue(true, options.getRequestCleanupInterval(),
-            sourceWriter == null ? null : sourceWriter.reconnectOutgoing);
         reconnectBufferSize = options.getReconnectBufferSize();
+
+        if (sourceWriter == null) {
+            normalOutgoing = new WriterMessageQueue(
+                options.getMaxMessagesInOutgoingQueue(),
+                options.isDiscardMessagesWhenOutgoingQueueFull(),
+                options.getRequestCleanupInterval());
+            reconnectOutgoing = new WriterMessageQueue();

```

Comment:
There are really 2 variants of the WriterMessageQueue, one for the normal user flow while connected and one for reconnecting, which only handles Protocol Messages and is never bounded. So I made different constructors.